### PR TITLE
Improve folding signature algorithm

### PR DESCRIPTION
**Pull request by ludusrusso**
Following #27, I've rewritten the folding algorithm in order to better fold headers and avoid introducing new lines inside headers. 

However, it seems it breaks the skim sign.

In my understanding, this should have been call after the signature was generated, but it seems that the new algorithm changes the generates signature itself. 

Any help?

### CODE REVIEW

**Comment by ludusrusso** [comment]
Example output:

```
        DKIM-Signature: a=ed25519-sha256;
         bh=2jUSOH9NhtVGCQWNr9BrIAPreKQjO6Sn7XIkfJVOzv8=; c=simple/simple;
         d=football.example.com; h=From:To:Subject:Date:Message-ID; s=brisbane;
         t=424242; v=1;
         b=CtvfpO96cCh7+PJLDna/bOJ1WdQ2qt7Nfrq6FzJqjYQv+TjOUc+rWk81TQJnmvdZsCDbmpr2
         hj7RDH4JDGa0CA==
```

**Comment by foxcpp** [comment]
>In my understanding, this should have been call after the signature was generated, but it seems that the new algorithm changes the generates signature itself.

The likely issue is that addition of non-empty `b=` changes how folding behaves. I tried to integrate go-message/textproto algorithm into go-msgauth and hit the exactly same issue.
You can observe it by adding fmt.Println for `sigField` here:
https://github.com/emersion/go-msgauth/blob/e4c87369d72f4a8d834a872ff0c43258936fd32d/dkim/sign.go#L261
and for `formatSignatrure(s.sigParams)` here:
https://github.com/emersion/go-msgauth/blob/e4c87369d72f4a8d834a872ff0c43258936fd32d/dkim/sign.go#L276

**Comment by foxcpp** [comment]
In your case, it shows:

```
DKIM-Signature: a=rsa-sha256;
 bh=2jUSOH9NhtVGCQWNr9BrIAPreKQjO6Sn7XIkfJVOzv8=; c=simple/simple;
 d=example.org; h=From:To:Subject:Date:Message-ID; s=brisbane; t=424242;
 v=1;
```
```
DKIM-Signature: a=rsa-sha256;
 bh=2jUSOH9NhtVGCQWNr9BrIAPreKQjO6Sn7XIkfJVOzv8=; c=simple/simple;
 d=example.org; h=From:To:Subject:Date:Message-ID; s=brisbane; t=424242;
 v=1;
 b=KXuOIdcHesiuc+CNVJC0ooTO9LvodnM01dstXNTA7elICw7r9LKD5zrPnGwtP5Ye+sOndhJO
 23pJIY3LP5oXzCnV3nkZiEwIb0rKIgYxgxx8JfCWbZgurhN35rQjnwCtmcGvN1aJ7ba1x1x63Vf
 xp0zwssLR2w8R+PkR4r1ROlM=
```

File: dkim/header.go [hunk 1]
```diff
@@ -92,37 +92,60 @@ func parseHeaderParams(s string) (map[string]string, error) {
...
+	var bvalue string
+	for k := range params {
+		if k == "b" {
+			bvalue = params["b"]
+		} else {
+			keys = append(keys, k)
+		}
+	}
+	sort.Strings(keys)
+	if found {
```
Comment (foxcpp) [inline dkim/header.go:143]: Never set to anything other than false.

File: dkim/header.go [hunk 2]
```diff
@@ -92,37 +92,60 @@ func parseHeaderParams(s string) (map[string]string, error) {
...
+			line = fmt.Sprintf("%v %v=%v;", line, k, v)
 		}
+	}
 
-		s += k + "=" + v + ";"
+	if line != "" {
+		s += line
 	}
+
+	if bvalue != "" {
```
Comment (foxcpp) [inline dkim/header.go:123]: Empty string is a valid value. The DKIM-Signature contents are signed with `b=`.

Comment (ludusrusso) [reply dkim/header.go:123]: Oh thanks, I've got it now! It seems it works :D

**Comment by ludusrusso** [comment]
> Additionally, I would appreciate if you add more test cases to ensure it generates valid signatures in edge cases related to folding.

Yea! I'll add some additional test cases! To be honest, I did not added more test before because they are covered by `TestSign`

**Comment by ludusrusso** [comment]
Now, if we have some part of the header longher than 75cahrs, the new algorithm will break the line length limit. This should not be an issues since this limit is only a recommendation! What do you think? https://github.com/emersion/go-msgauth/issues/18#issuecomment-597186179

Anyway this should fix #18

**Comment by foxcpp** [comment]
I agree that 75 chars limitation can be ignored. Following rationale in RFC, it is not really relevant today as a rare person looks at message header.

**Comment by ludusrusso** [comment]
@foxcpp thanks! I've also added a more test that checks only if the headers names in the "h" values are properly folded in the generated signature!

**Comment by foxcpp** [comment]
FWIW, That's the h= value my mail server used for a typical plaintext message:
```
Subject:Subject:Sender:To:To:Cc:From:From:Date:Date:MIME-Version:Content-Type:Content-Type:Content-Transfer-Encoding:Reply-To:In-Reply-To:Message-Id:Message-Id:References:Autocrypt:Openpgp
```

Interesting is that old folding algorithm broke it in the middle of a field name yet
OpenDKIM verified the signature successfully:
https://paste.dn42.us/paste/#/8T1GgumCllH0ZGDbwvOI8dvNYA8!TqXFCKZWbnYkBUP4_rBv1Fd3e-OVScQBZDav2mXSMw4

**Comment by foxcpp** [comment]
@emersion, what do you think about this PR?

**Comment by ludusrusso** [comment]
> FWIW, That's the h= value my mail server used for a typical plaintext message:
> 
> ```
> Subject:Subject:Sender:To:To:Cc:From:From:Date:Date:MIME-Version:Content-Type:Content-Type:Content-Transfer-Encoding:Reply-To:In-Reply-To:Message-Id:Message-Id:References:Autocrypt:Openpgp
> ```
> 
> Interesting is that old folding algorithm broke it in the middle of a field name yet
> OpenDKIM verified the signature successfully:
> https://paste.dn42.us/paste/#/8T1GgumCllH0ZGDbwvOI8dvNYA8!TqXFCKZWbnYkBUP4_rBv1Fd3e-OVScQBZDav2mXSMw4

I also notice this! I've tried with different mail provider receivers and in some case (like hotmail/outlook) the dkim fails while other (like gmail) it passes.

File: dkim/header.go [hunk 3]
```diff
@@ -92,37 +92,58 @@ func parseHeaderParams(s string) (map[string]string, error) {
...
+	var line string
 	first := true
+
 	for _, k := range keys {
 		v := params[k]
 
 		if first {
 			first = false
+			line += fmt.Sprintf("%v: %v=%v;", headerFieldName, k, v)
+			continue
```
Comment (emersion) [inline dkim/header.go:108]: This `continue` statement is unnecessary

File: dkim/header.go [hunk 4]
```diff
@@ -92,37 +92,58 @@ func parseHeaderParams(s string) (map[string]string, error) {
...
 	var s string
+	var line string
 	first := true
+
 	for _, k := range keys {
 		v := params[k]
 
 		if first {
 			first = false
+			line += fmt.Sprintf("%v: %v=%v;", headerFieldName, k, v)
```
Comment (emersion) [inline dkim/header.go:107]: We could initialize `s` like so:

```go
s := headerFieldName + ":"
```

Then we wouldn't need `first` anymore.

File: dkim/header.go [hunk 5]
```diff
@@ -92,37 +92,58 @@ func parseHeaderParams(s string) (map[string]string, error) {
...
-		} else {
-			keys = append(keys, k)
-		}
-	}
-	sort.Strings(keys)
-	if found {
-		keys = append(keys, "b")
-	}
+func formatHeaderParams(headerFieldName string, params map[string]string) string {
+	keys, bvalue, found := sortParams(params)
```
Comment (emersion) [inline dkim/header.go:96]: Can we rename `found` to `bfound`, to be a little more explicit?

**Comment by ludusrusso** [comment]
Hi @emersion, thanks for the feedback! I was wondering: if we do not want to limit the line lenght to 75 chars, we could make the folding algorithm simpler by just putting the "b" value at the end of the header without folding it! What do you think about this?

**Comment by emersion** [comment]
Would the `b=` value ever be longer than 1000 chars? (I guess that's an issue with other header fields too, like `h=`)

**Comment by ludusrusso** [comment]
> Would the `b=` value ever be longer than 1000 chars? (I guess that's an issue with other header fields too, like `h=`)

Not super sure about `b`, but since it's an hash it should be fixed in length, I'm correct?

Regarding `h`, you're right, it could be more then 1000 chars if you include a lot of  headers! I've to think about a way to fold it :D

**Comment by ludusrusso** [comment]
@emersion what do you think about b folding?

**Comment by foxcpp** [comment]
`b=` is a signature value. I do not expect it to hit 998 characters limitation even with RSA-4096 (though one might consider it to be dangerously close, ~682 characters in base64)
